### ISSUE TITLE
License prevents commercial usage of RedisGraph

### ISSUE THREAD

**Issue by gkorland** [FalkorDB/FalkorDB#442]
Created by: @bionicles
Source: https://github.com/RedisGraph/RedisGraph/issues/3098
I’d love to use this software for actual business projects, but unfortunately, neither RSALv2 nor SSPL are viable, so the code is essentially useless since you can’t make a business app using RedisGraph on either license :( can you ask your legal team to wake up, quit being cowards in the competition with cloud service providers, and relicense the software under a dual MIT/Apache 2.0 license? Then we could actually use this code for work. It’s really cool what you’ve done and I want to use it. I just can’t, for legal reasons, because your weird licensing scheme prohibits it. Since it’s not commercially viable to use RedisGraph, I have zero reason to pay money to your employer.

Hope I makes sense: I like what you have done here and I want to use it, but I can’t, which means your legal team is trying to write fancy language to make more money, which prevents me from actually paying you money. 

RSALv2 is completely ridiculous:
“””
Limitations

You may not make the functionality of the software or a modified version available to third parties as a service, or distribute the software or a modified version in a manner that makes the functionality of the software available to third parties.

Making the functionality of the software or modified version available to third parties includes, without limitation, enabling third parties to interact with the functionality of the software or modified version in distributed form or remotely through a computer network, offering a product or service the value of which entirely or primarily derives from the value of the software or modified version, or offering a product or service that accomplishes for users the primary purpose of the software or modified version.

You may not alter, remove, or obscure any licensing, copyright, or other notices of the licensor in the software. Any use of the licensor’s trademarks is subject to applicable law.
“””
TLDR: you can’t let customers use your database if you build with RedisGraph. What is the point of the project if customers can’t actually use it?

From SSPL, which looks good until this part:
“””
13. Offering the Program as a Service.

If you make the functionality of the Program or a modified version available to third parties as a service, you must make the Service Source Code available via network download to everyone at no charge, under the terms of this License. Making the functionality of the Program or modified version available to third parties as a service includes, without limitation, enabling third parties to interact with the functionality of the Program or modified version remotely through a computer network, offering a service the value of which entirely or primarily derives from the value of the Program or modified version, or offering a service that accomplishes for users the primary purpose of the Program or modified version.

“Service Source Code” means the Corresponding Source for the Program or the modified version, and the Corresponding Source for all programs that you use to make the Program or modified version available as a service, including, without limitation, management software, user interfaces, application program interfaces, automation software, monitoring software, backup software, storage software and hosting software, all such that a user could run an instance of the service using the Service Source Code you make available.
“””
Basically, you aren’t allowed to have private code which uses RedisGraph. Not all code is suitable for publishing. Not all code is safe to publish.

What is the point of RedisGraph being a great program, if we can’t use it for work stuff?

**Comment by gkorland** [comment]
Origin comment by: @bionicles
Here’s an example of a business friendly permissive open source license: https://github.com/bionicles/tree_plus/blob/main/LICENSE

**Comment by gkorland** [comment]
Origin comment by: @bionicles
Just feel frustrated because I love Redis and RedisGraph and I don’t want to wait around for PostgreSQL to support Cypher (which is now part of SQL, I might add, thus dramatically reducing the need for RedisGraph)

**Comment by gkorland** [comment]
Origin comment by: @LiorKogan
If you want to include Redis Enterprise or a Redis RSALv2/SSPLv1 module as part of your commercial product you should check our partner program. You won't be the first one who's doing that.

https://redis.com/press/redis-announces-enhanced-global-partner-program/

**Comment by gkorland** [comment]
Origin comment by: @gkorland
@bionicles can you please share more information about your product? 
RSALv2 should not limit your usage unless your product is a service:
```
offering a product or service the value of which entirely or primarily derives from the value of the software or modified version, or offering a product or service that accomplishes for users the primary purpose of the software or modified version.
```

There are many commercial companies that are using RedisGraph free source available version in production without violating the license.

**Comment by gkorland** [comment]
Origin comment by: @bionicles
thanks for the reply @gkorland! Can you list examples of those commercial companies that are using RedisGraph free source available version in production without violating the license? That might help put the issue to rest. 

Anyway, I'm a huge proponent of on-demand autoscaling pricing, so it's definitely a service version of an open source tool and not a product one might buy once or whatever. 

For the record, it was never and is not my intention to just host RedisGraph for people and give them an api endpoint to a graph database, but rather to use RedisGraph as the backend of various diverse projects: for example I made an algotrading AI system, a molecular optimization system, and a general AI system, among other projects.

The line between product and service is really fuzzy. I'm sure I could use it and just trust you guys not to sue me or whatever, but that sounds like a pretty silly strategy. Hope is not a strategy. If it's even a minor risk of breaking your terms, I just go to one of the bajillion other databases with normal permissive licenses (i.e. MIT & Apache 2.0), like PostgreSQL or ArangoDB, for example.

let 'software' mean RedisGraph here:
I'll try to clarify the definition of "entirely or primarily derives from the value of the software." 

To clarify why "entirely or primarily derives from the value of the software." concerns me and i consider it to be an extremely dangerous statement, I'll give you a single word:
**counterfactuals**

To apply counterfactuals to this issue, I'll use the language pattern "**if not X, then Y, because Z**"

**If not for their use of a database, then those commercial companies would be worthless, because they would not have a system of record for their paying customers, which means the customers might as well run whatever thing it is locally and pay nothing, especially given the context of graph data in which the valuable thing is the linkages between entities.** 

Thus, I argue, the only way "entirely or primarily derives from the value of the software" doesn't apply to a commercial company using RedisGraph, is if you literally use a different database for over 50% of your data (ignoring the nonlinearities involved, let's assume primarily means 'over 50%'), and RedisGraph is just an optional nice to have, in which case the counterfactual becomes

**If not for the use of RedisGraph as a database, then those commercial companies would still be valuable, because they don't rely on RedisGraph for the entirety or 'primary bulk' of their data.** 

Would Twitter be valuable without their data? No. Would Facebook? No. How about Google or OpenAI? No, 

To say "you can use RedisGraph as long as you don't derive value from it" basically implies, "don't use this to create value or it breaks the rules!"

Aka, the terms of the license imply (to my non lawyer engineer brain anyway) that the only way you're allowed to use RedisGraph is if it's **_optional_** , in which case, Redis Source Available License (v2) implies the only way you're allowed to use RedisGraph for business, is if the software is, for lack of a better term, **cruft.** 

:(

**Comment by gkorland** [comment]
Origin comment by: @hpvd
RSAL grants equivalent rights to permissive open source licenses for the vast majority of users. With RSAL,
- developers can use the software; modify the source code’ **integrate it with an application**; and use, distribute or sell their application.
- The only restriction is that **the application cannot be a** database, a caching engine, a stream processing engine, a search engine, an indexing engine or an ML/DL/AI serving engine.

source:
https://redis.com/blog/redis-labs-modules-license-changes/

to me, that looks like
- **you can** e.g. build an **ecommerce app** build on redis and use or distribute or sell it
- **you can not** take redisgraph name it bionicelsgraph and sell it as a **new database**

@gkorland could you confirm this?

Since this topic is really confusing for every existing or even new user thinking about using redisgraph, it would be good to solve it as fast and complete as possible.

**Comment by gkorland** [comment]
Origin comment by: @hpvd
ah ok, redis graph is EOL now.
So one do not have to worry about the license, since there is now longterm bright future to build on...
see https://redis.com/blog/redisgraph-eol/

**Comment by gkorland** [comment]
Origin comment by: @hpvd
moving on to terminusdb which looks pretty interesting as git like (version control) in memory graph db under Apache 2.0 License
https://github.com/terminusdb/terminusdb

see https://terminusdb.com/blog/9-reasons-why-you-need-an-immutable-database/

**Comment by gkorland** [comment]
Origin comment by: @gkorland
The team behind RedisGraph decided to take a stab at bringing this disruptive technology back to market.
Check out - https://github.com/falkordb/falkordb

**Comment by hpvd** [comment]
great to see this tec lives on as falkorDB :-)

regarding licences:

@gkorland can you confirm this for falkorDBs licence?

> 
> to me, that looks like
> 
> - you can e.g. build an ecommerce app build on falkorDB and use or distribute or sell it
> - you can not take falkorDB name it `oneandonlyrealgraph` and sell it as a new database

 
 if so, maybe its worth a short blogpost...

**Comment by bionicles** [comment]
"The only restriction is that the application cannot be a database, a caching engine, a stream processing engine, a search engine, an indexing engine or an ML/DL/AI serving engine."

Is this a JOKE?

**Comment by hpvd** [comment]
specified ecommerce app features in post above

**Comment by gkorland** [comment]
FalkorDB is [SSPL](https://raw.githubusercontent.com/FalkorDB/FalkorDB/master/LICENSE.txt) like many of the products today like MongoDB and ElasticSearch
